### DISCUSSION TITLE
Feedback thread for architecture proposal

### DISCUSSION THREAD

**Discussion by jeremyckahn** [passway-project/passway#2] [General]
Please see: https://github.com/passway-project/passway/wiki/Passway-architecture

And please share your feedback in this thread!

**Comment by thatsnotright** [comment]
Have you considered using S3 or minio as your object store (they already support chunk based streaming) along side a simple sqlite database for each user? This would let you track any associated metadata with the blobs in a database (which can optionally be encrypted per user key) and you'd maybe just need a master sqlite that associates the user with their main sqlite database.

The reasons I like this approach over mongo is that you don't have to fight with mongo doing something it isn't very good at (which is most things, frankly), you can securely store metadata about user uploads per user and have all of that metadata encrypted on the client before being saved (sqlite has been compiled to wasm). Then the only database which has any information that's potentially unencrypted by the user is a main database linking a root object path with a user.

**Reply by jeremyckahn** [reply]
> Have you considered using S3 or minio as your object store

> The reasons I like this approach over mongo is that you don't have to fight with mongo doing something it isn't very good at

S3 was actually my initial choice for this, but I thought about it some more and realized that it goes against the self-hosting ethos of the project. I think there's room for using it as an alternative back end in a later iteration (or any other object store that folks come up with an adapter for), but I'd prefer to not use it for the initial iteration.

I haven't actually used MongoDB myself yet, so my opinions about it are pretty naive. Do you know of another open source, self-hostable solution that might be a better fit here? I'd like to avoid technology choices that would lock users into a particular cloud vendor or platform.

> maybe just need a master sqlite that associates the user with their main sqlite database.

This is an interesting idea. I was on the fence about whether the user metadata and content should live in the same store (as I currently have it). I'm open to the idea of separating them and storing user metadata in a SQL DB of some kind. My first thought would be to use PostGres, though. @thatsnotright do you think SQLite might be a better fit for this?

> you can securely store metadata about user uploads per user and have all of that metadata encrypted on the client before being saved (sqlite has been compiled to wasm). Then the only database which has any information that's potentially unencrypted by the user is a main database linking a root object path with a user.

I'm not totally following this idea. I don't think we want to end-to-end encrypt the metadata because then the server doesn't know how to look anything up (such as content IDs and sizes) to provide back to users later. It's also not clear how the server might be able to do periodic data evictions if content metadata is encrypted. Could you expand on how you see that potentially working?

**Reply by thatsnotright** [reply]
> I haven't actually used MongoDB myself yet, so my opinions about it are pretty naive. Do you know of another open source, self-hostable solution that might be a better fit here? I'd like to avoid technology choices that would lock users into a particular cloud vendor or platform.

minio fits this perfectly, then, without messing with mongo. Personally, I would avoid mongo at all costs. and it's s3 compatible for users who can spread their data across s3 urls or minio urls (or file system urls).

https://github.com/minio/minio
 
> This is an interesting idea. I was on the fence about whether the user metadata and content should live in the same store (as I currently have it). I'm open to the idea of separating them and storing user metadata in a SQL DB of some kind. My first thought would be to use PostGres, though. @thatsnotright do you think SQLite might be a better fit for this?
> 
> > you can securely store metadata about user uploads per user and have all of that metadata encrypted on the client before being saved (sqlite has been compiled to wasm). Then the only database which has any information that's potentially unencrypted by the user is a main database linking a root object path with a user.
> 
> I'm not totally following this idea. I don't think we want to end-to-end encrypt the metadata because then the server doesn't know how to look anything up (such as content IDs and sizes) to provide back to users later. It's also not clear how the server might be able to do periodic data evictions if content metadata is encrypted. Could you expand on how you see that potentially working?

what I am proposing is more like a federated store, where only data relevant to that level is stored at that level (which is why I suggested sqlite). This way, there could be only one main database storing all the associations from a user to other sqlite databases which are specific for each user and stored elsewhere (minio, some other s3 bucket, a raw file system, whatever can support a url schema and the client can understand how to access). This allows distributed access and only needs a little bit of work for concurrent updates to sqlite but since it's per user the conflicts only exist for that user (and you could make the highest level database a hosted postgres instance if you're worried about write conflicts at that level).

you can also encrypt the sqlite database such that two users can decrypt it but at rest it's in a form not readable by others so any compromise would need to get the service key or the user's key to decrypt the metadata for that user.

storing a date/time to eject content could be done without knowing how to describe those blobs, just keep the url to eject(delete) and a timestamp, then a 404 on that by the user would indicate the data has been removed. or if you encrypt with 2 keys you can update the users local sqlite store.

regardless, if the user metadata were stored in a sqlite file then the server would only really be responsible for retrieving that blob, keeping some kind of 'checked out' status to understand if users are going to make writes, and sending it to the user client. the client would decrypt it and process the entries. it'd act more like a directory than a service, with clients needing to be smarter/do more heavy lifting to support the federated/distributed nature once they receive the sqlite db from the directory service.

**Reply by jeremyckahn** [reply]
Thanks for the explanation, @thatsnotright. I have to admit that I'm still struggling to understand the idea being presented here, so I don't feel comfortable building on top of it.

It sounds like you're advocating for a more decentralized or federated system. Decentralization and federation have their place (and I'm generally a fan of htem), but that's not a goal for this project. I specifically want to build a centralized system, because that's simplest way to achieve a highly available and performant service.

----------

That said, I will look more into Minio. I initially thought that it was an AI-specific data store, but it appears to have more use cases that may make sense for Passway. Thank you for the recommendation!

**Reply by jeremyckahn** [reply]
I spent some time looking into MinIO as an alternative to MongoDB. Overall, I think they'd both be good options for what Passway needs. Given that the user content data store will be accessed via an arbitrating web app, the choice of which DB to use is mostly an implementation detail that should be abstracted away by a sensible interface. This will allow users to swap one DB out for another if they choose.

Between MinIO and MongoDB, MinIO seems the more lightweight option and therefore more appealing to start with. Thanks for the suggestion @thatsnotright!

**Comment by thatsnotright** [comment]
Just to clarify, the system doesn't have to be decentralized but it would be easier to add it if someone desired that behavior. It also would reduce risk of a catastrophic data failure on your main database, since each user would get their own index database rebuilding the loss of the main database would just involve walking the tree of objects and processing each user database.

**Reply by jeremyckahn** [reply]
> It also would reduce risk of a catastrophic data failure on your main database

It seems best to solve this problem with centrally-managed server redundancy and other resiliency measures. There's only so much complexity that should be pushed to the client for this project, as doing so makes system modification and management much more difficult. I've recently updated the architecture design to be simpler, and at this point it's not much more complex than your average CRUD app. I think this simplicity is a significant benefit to the project, and it seems best to avoid deviating from that any more than necessary.

---------------

There is definitely a place for decentralized systems and mesh networks (and I am big fan of them!), but I don't think it's a fit for what this project is aiming to achieve. I believe centralization should be a goal of Passway in order to optimize for both performance and reliability.
